pkg/controller/node/nodecontroller_test.go

```diff
@@ -458,191 +457,689 @@ func TestMonitorNodeStatusEvictPods(t *testing.T) {
+						},
+					},
+				},
+				{
+					Conditions: []api.NodeCondition{
+						{
+							Type:   api.NodeReady,
+							Status: api.ConditionUnknown,
+							// Node status was updated by nodecontroller 1hr ago
+							LastHeartbeatTime:  unversioned.Date(2015, 1, 1, 12, 0, 0, 0, time.UTC),
+							LastTransitionTime: unversioned.Date(2015, 1, 1, 12, 0, 0, 0, time.UTC),
+						},
+					},
+				},
+				healthyNodeNewStatus,
+				healthyNodeNewStatus,
+			},
+			expectedInitialStates: map[string]zoneState{
+				createZoneID("region1", "zone1"): statePartialDisruption,
+			},
+			expectedFollowingStates: map[string]zoneState{
+				createZoneID("region1", "zone1"): statePartialDisruption,
+			},
+			expectedEvictPods: true,
+			description:       "Zone is partially disrupted - eviction should take place.",
 		},
 	}
 
 	for _, item := range table {
-		nodeController, _ := NewNodeController(nil, item.fakeNodeHandler,
+		fakeNodeHandler := &FakeNodeHandler{
+			Existing:  item.nodeList,
+			Clientset: fake.NewSimpleClientset(&api.PodList{Items: item.podList}),
+		}
+		nodeController, _ := NewNodeController(nil, fakeNodeHandler,
 			evictionTimeout, testRateLimiterQPS, testNodeMonitorGracePeriod,
 			testNodeStartupGracePeriod, testNodeMonitorPeriod, nil, nil, 0, false)
 		nodeController.now = func() unversioned.Time { return fakeNow }
+		nodeController.enterPartialDisruptionFunc = func(nodeNum int, defaultQPS float32) float32 {
+			return testRateLimiterQPS
+		}
+		nodeController.enterFullDisruptionFunc = func(nodeNum int, defaultQPS float32) float32 {
+			return testRateLimiterQPS
+		}
 		for _, ds := range item.daemonSets {
 			nodeController.daemonSetStore.Add(&ds)
 		}
 		if err := nodeController.monitorNodeStatus(); err != nil {
-			t.Errorf("unexpected error: %v", err)
+			t.Errorf("%v: unexpected error: %v", item.description, err)
 		}
-		if item.timeToPass > 0 {
-			nodeController.now = func() unversioned.Time { return unversioned.Time{Time: fakeNow.Add(item.timeToPass)} }
-			item.fakeNodeHandler.Existing[0].Status = item.newNodeStatus
-			item.fakeNodeHandler.Existing[1].Status = item.secondNodeNewStatus
+
+		for zone, state := range item.expectedInitialStates {
+			if state != nodeController.zoneStates[zone] {
+				t.Errorf("%v: Unexpected zone state: %v: %v instead %v", item.description, zone, nodeController.zoneStates[zone], state)
+			}
+		}
+
+		nodeController.now = func() unversioned.Time { return unversioned.Time{Time: fakeNow.Add(timeToPass)} }
+		for i := range item.updatedNodeStatuses {
+			fakeNodeHandler.Existing[i].Status = item.updatedNodeStatuses[i]
 		}
+
 		if err := nodeController.monitorNodeStatus(); err != nil {
-			t.Errorf("unexpected error: %v", err)
+			t.Errorf("%v: unexpected error: %v", item.description, err)
 		}
-		zones := getZones(item.fakeNodeHandler)
+		// Give some time for rate-limiter to reload
+		time.Sleep(50 * time.Millisecond)
```

Comment:
When we swap the rate limiter it's drained at the begining, so we need to wait for it to regain a token.